analysis/src/vm/romgen.rs

```diff
@@ -0,0 +1,217 @@
+                    .map(|r| format!("{}' = 0", r.name))
+                    .collect::<Vec<_>>()
+                    .join(", ")
+            )),
+            parse_instruction_definition(&format!("instr _loop {{ {}' = {} }}", pc, pc)),
+        ];
+
+        machine.instructions.extend(embedded_instructions);
+
+        // implement the return instruction. For this dispatcher, returning jumps to 0
+        machine
+            .instructions
+            .iter_mut()
+            .find(|i| i.name == "return")
+            .unwrap()
+            .body = parse_instruction_body(&format!("{{ {}' = 0 }}", machine.pc().unwrap()));
+
+        // generate the rom
+        // the functions are already batched, we just batch the dispatcher manually here
+        // we could also run batching again on the dispatcher
+
+        let mut rom: Vec<Batch<T>> = vec![];
+
+        // add the beginning of the dispatcher
+        rom.extend(vec![
+            Batch::from(vec![
+                parse_function_statement("_start::"),
+                parse_function_statement("_reset;"),
+            ])
+            .reason(IncompatibleSet::from(Incompatible::Unimplemented)),
+            // the latch is constrained to start at 0 using _L1
+            Batch::from(vec![parse_function_statement("_jump_to_operation;")])
+                .reason(IncompatibleSet::from(Incompatible::Label)),
+        ]);
+
+        // add each function, setting the function_id to the current position in the ROM
+        for function in machine.functions.iter_mut() {
+            function.id = Some(T::from(rom.len() as u64));
+
+            let mut batches: Vec<_> = function
+                .body
+                .statements
+                .clone()
+                .into_iter_batches()
+                .collect();
+            // modify the first batch to include the label just for debugging purposes, it's always possible to batch it so it's free
+            batches
+                .first_mut()
+                .expect("function should have at least one statement as it must return")
+                .statements
+                .insert(
+                    0,
+                    parse_function_statement(&format!("_{}::", function.name)),
+                );
+
+            // modify the last batch to be caused by the coming label
+            let last = batches
+                .last_mut()
+                .expect("function should have at least one statement as it must return");
+            last.set_reason(IncompatibleSet::from(Incompatible::Label));
+
+            rom.extend(batches);
+        }
+
+        // THE FOLLOWING IS NECESSARY BECAUSE OF WITGEN, IT COULD BE REMOVED ONCE WITGEN CAN CALL THE INFINITE LOOP ITSELF
+        // we get the location of the sink so that witgen jumps to it after the first call is done
+        // this constrains the VM to being able to execute only one call, which will be fixed in the future
+        let sink_id = T::from(rom.len() as u64);
```

Comment:
this value is also used by the loop above, could be declared before

Reply:
No, this value changes as we add statements to the rom. Related to your question above.